homeassistant/components/tellduslive/__init__.py

```diff
@@ -230,24 +139,37 @@ def identify_device(device):
                 "Unidentified device type (methods: %d)", device.methods)
             return 'switch'
 
-        def discover(device_id, component):
+        async def discover(device_id, component):
             """Discover the component."""
-            discovery.load_platform(
-                self._hass, component, DOMAIN, [device_id], self._config)
+            config_entries_key = '{}.{}'.format(component, DOMAIN)
+            async with self._hass.data[DATA_CONFIG_ENTRY_LOCK]:
+                if config_entries_key not in self._hass.data[
+                        CONFIG_ENTRY_IS_SETUP]:
+                    await self._hass.config_entries.async_forward_entry_setup(
+                        self._config, component)
+                    self._hass.data[CONFIG_ENTRY_IS_SETUP].add(
+                        config_entries_key)
+
+            async_dispatcher_send(
+                self._hass, TELLDUS_DISCOVERY_NEW.format(component, DOMAIN),
+                device_id)
 
         for device in self._client.devices:
             if device.device_id in self._known_devices:
                 continue
             if device.is_sensor:
                 for item in device.items:
-                    discover((device.device_id, item.name, item.scale),
-                             'sensor')
+                    await discover((device.device_id, item.name, item.scale),
```

Comment:
The important order is to call `device.info` before `discover` but you are right we could call `async_dispatcher_send` in bulk once all `device.info` is done for that type. I'm not sure how much benefit there will be though as it will probably only happen in bulk during initial load.

Reply:
We generally should avoid awaiting in loops when order is not important, since that defeats a major win of asyncio.